smb/smb.go

```diff
@@ -754,6 +816,554 @@ type QueryDirectoryRes struct {
+	if err != nil {
+		log.Errorln(err)
+		return nil, err
+	}
+	err = binary.Write(w, binary.LittleEndian, sidBuf)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+	return w.Bytes(), nil
+}
+
+func readACE(r *bytes.Reader) (a *ACE, err error) {
+	a = &ACE{}
+	err = binary.Read(r, binary.LittleEndian, &a.Header.Type)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+
+	err = binary.Read(r, binary.LittleEndian, &a.Header.Flags)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+
+	err = binary.Read(r, binary.LittleEndian, &a.Header.Size)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+
+	err = binary.Read(r, binary.LittleEndian, &a.Mask)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+
+	sid, err := readSID(r)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+	a.Sid = *sid
+
+	return
+}
+
+func (self *ACE) UnmarshalBinary(buf []byte) (err error) {
+	r := bytes.NewReader(buf)
+	ace, err := readACE(r)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+
+	*self = *ace
+	return nil
+}
+
+func (self *PACL) MarshalBinary() (ret []byte, err error) {
+	w := bytes.NewBuffer(ret)
+
+	err = binary.Write(w, binary.LittleEndian, self.AclRevision)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+	err = binary.Write(w, binary.LittleEndian, self.AclSize)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+
+	// Encode AceCount at 4 byte boundary
+	err = binary.Write(w, binary.LittleEndian, uint32(len(self.ACLS)))
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+
+	for _, item := range self.ACLS {
+		var aceBuf []byte
+		aceBuf, err = item.MarshalBinary()
+		if err != nil {
+			log.Errorln(err)
+			return
+		}
+		_, err = w.Write(aceBuf)
+		if err != nil {
+			log.Errorln(err)
+			return
+		}
+	}
+
+	return w.Bytes(), nil
+}
+
+func readPACL(r *bytes.Reader) (p *PACL, err error) {
+	p = &PACL{}
+	err = binary.Read(r, binary.LittleEndian, &p.AclRevision)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+
+	err = binary.Read(r, binary.LittleEndian, &p.AclSize)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+	err = binary.Read(r, binary.LittleEndian, &p.AceCount)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+
+	p.ACLS = make([]ACE, p.AceCount)
+	for i := range p.ACLS {
+		var ace *ACE
+		ace, err = readACE(r)
+		if err != nil {
+			log.Errorln(err)
+			return
+		}
+		p.ACLS[i] = *ace
+	}
+
+	return
+}
+
+func (self *PACL) UnmarshalBinary(buf []byte) (err error) {
+	r := bytes.NewReader(buf)
+	pacl, err := readPACL(r)
+	if err != nil {
+		log.Errorln(err)
+		return
+	}
+	*self = *pacl
+
+	return nil
+}
+
+var (
+	accessMaskMap = map[uint32]string{
+		0x80000000: "GENERIC_READ",
+		0x4000000:  "GENERIC_WRITE",
+		0x20000000: "GENERIC_EXECUTE",
+		0x10000000: "GENERIC_ALL",
+		0x02000000: "MAXIMUM_ALLOWED",
+		0x01000000: "ACCESS_SYSTEM_SECURITY",
+		0x00100000: "SYNCHRONIZE",
+		0x00080000: "WRITE_OWNER",
+		0x00040000: "WRITE_DACL",
+		0x00020000: "READ_CONTROL",
+		0x00010000: "DELETE",
+	}
+)
```

Comment:
Move map to earlier in the file with the rest of the variable and constant declarations, before the beginning of struct definitions